src/pypyjs.js

```diff
@@ -0,0 +1,1103 @@
+      const xhr = new XMLHttpRequest();
+      xhr.onload = function onload() {
+        if (xhr.status >= 400) {
+          reject(xhr);
+        } else {
+          resolve(xhr);
+        }
+      };
+
+      xhr.open('GET', rootURL + relpath, true);
+      xhr.responseType = responseType || 'text';
+      xhr.send(null);
+    });
+  }
+
+  // For nodejs, use fs.readFile.
+  if (typeof fs !== 'undefined' && typeof fs.readFile !== 'undefined') {
+    return new Promise((resolve, reject) => {
+      fs.readFile(path.join(rootURL, relpath), (err, data) => {
+        if (err) return reject(err);
+        resolve({ responseText: data.toString() });
+      });
+    });
+  }
+
+  // For spidermonkey, use snarf (which has a binary read mode).
+  if (typeof snarf !== 'undefined') {
+    return new Promise((resolve) => {
+      const data = snarf(rootURL + relpath);
+      resolve({ responseText: data });
+    });
+  }
+
+  // For d8, use read() and readbuffer().
+  if (typeof read !== 'undefined' && typeof readbuffer !== 'undefined') {
+    return new Promise((resolve) => {
+      const data = read(rootURL + relpath);
+      resolve({ responseText: data });
+    });
+  }
+
+  return new Promise((resolve, reject) => {
+    reject('unable to fetch files');
+  });
+};
+
+// add a Module to the vm from a file so you can import it later. Uses a path
+// relative to pypyjs.
+pypyjs.prototype.addModuleFromFile = function addModuleFromFile(name, file) {
+  return this.fetch(file).then((data) => this.addModule(name, data.responseText));
+};
+
+// add a Module to the vm from javascript. It will automatically write it it to
+// the "disk" and lets you import it in your code later.
+pypyjs.prototype.addModule = function addModule(name, source) {
+  return this.findImportedNames(source).then((imports) => {
+    // keep track of any modules that have been previously loaded
+    if (this._loadedModules[name]) {
+      this._modulesToReset[name] = true;
+      this._loadedModules[name] = null;
+    }
+    this._allModules[name] = {
+      file: `${name}.py`,
+      imports
+    };
+    if (!this.inJsModules) {
+      this.inJsModules = [];
+    }
+    this.inJsModules[`modules/${name}.py`] = source;
+  });
+};
+
+function _blockIndent(code, indent) {
+  return code.replace(/\n/g, `\n${indent}`);
+}
+
+function _escape(value) {
+  return value.replace(/\\/g, '\\\\').replace(/'/g, '\\\'');
+}
+
+// Method to execute python source directly in the VM.
+//
+// This is the basic way to push code into the pypyjs VM.
+// Calling code should not use it directly; rather we use it
+// as a primitive to build up a nicer execution API.
+//
+pypyjs.prototype._execute_source = function _execute_source(code, preCode) {
+  const Module = this._module;
+  const _preCode = preCode ? preCode : '';
+  let code_ptr;
+
+  return new Promise(function promise(resolve, reject) {
+    const _code = `try:
+  ${_blockIndent(_preCode, '  ')}
+  ${code}
```

Comment:
Ah, gotcha.  It might be clearer as a separate function, or an explicit "don't use top-level scope" option.  Does ES6 offer anything fun like named function arguments?  If so we could do something like:

```
_execute_source(preCode, unscoped=true).then(function () {
  return _execute_source(code)
})
```

Rather than combining the two into a single code string.